pkg/uhttp/dbcache.go

```diff
@@ -0,0 +1,674 @@
+				l.Debug(failRollback, zap.Error(errtx))
+			}
+
+			l.Debug(failInsert, zap.Error(err))
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Has query for cached keys.
+func (d *DBCache) Has(ctx context.Context, key string) (bool, error) {
+	if d.IsNilConnection() {
+		return false, fmt.Errorf("%s", nilConnection)
+	}
+
+	l := ctxzap.Extract(ctx)
+	rows, err := d.db.Query("SELECT data FROM http_cache where key = ?", key)
+	if err != nil {
+		l.Debug("Failed to query cache table", zap.Error(err))
+		return false, err
+	}
+
+	defer rows.Close()
+	for rows.Next() {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// IsNilConnection check if the database connection is nil.
+func (d *DBCache) IsNilConnection() bool {
+	return d.db == nil
+}
+
+// Select query for cached response.
+func (d *DBCache) Select(ctx context.Context, key string) ([]byte, error) {
+	var data []byte
+	if d.IsNilConnection() {
+		return nil, fmt.Errorf("%s", nilConnection)
+	}
+
+	l := ctxzap.Extract(ctx)
+	rows, err := d.db.QueryContext(ctx, "SELECT data FROM http_cache where key = ?", key)
+	if err != nil {
+		l.Debug(errQueryingTable, zap.Error(err))
+		return nil, err
+	}
+
+	defer rows.Close()
+	for rows.Next() {
+		err = rows.Scan(&data)
+		if err != nil {
+			l.Debug("Failed to scan rows from cache table", zap.Error(err))
+			return nil, err
+		}
+	}
+
+	return data, nil
+}
+
+func (d *DBCache) Remove(ctx context.Context, key string) error {
+	if d.IsNilConnection() {
+		return fmt.Errorf("%s", nilConnection)
+	}
+
+	l := ctxzap.Extract(ctx)
+	tx, err := d.db.Begin()
+	if err != nil {
+		l.Debug(failStartTransaction, zap.Error(err))
+		return err
+	}
+
+	_, err = d.db.ExecContext(ctx, "DELETE FROM http_cache WHERE key = ?", key)
+	if err != nil {
+		if errtx := tx.Rollback(); errtx != nil {
+			l.Debug(failRollback, zap.Error(errtx))
+		}
+
+		l.Debug("error deleting key", zap.Error(err))
+		return err
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		if errtx := tx.Rollback(); errtx != nil {
+			l.Debug(failRollback, zap.Error(errtx))
+		}
+
+		l.Debug("Failed to remove cache value", zap.Error(err))
+		return err
+	}
+
+	return nil
+}
+
+func (d *DBCache) close(ctx context.Context) error {
+	if d.IsNilConnection() {
+		return fmt.Errorf("%s", nilConnection)
+	}
+
+	err := d.db.Close()
+	if err != nil {
+		ctxzap.Extract(ctx).Debug("error closing database", zap.Error(err))
+		return err
+	}
+
+	return nil
+}
+
+// Expired checks if key is expired.
+func (d *DBCache) Expired(expiration int64) bool {
+	return time.Now().UnixNano() > expiration
+}
+
+// Delete all expired items from the cache.
+func (d *DBCache) deleteExpired(ctx context.Context) error {
+	var (
+		expiration     int64
+		key            string
+		mapExpiredKeys = make(map[string]bool)
+	)
+
+	if d.IsNilConnection() {
+		return fmt.Errorf("%s", nilConnection)
+	}
+
+	l := ctxzap.Extract(ctx)
+	rows, err := d.db.QueryContext(ctx, "SELECT key, expiration FROM http_cache")
```

Comment:
Instead of iterating over every key in the cache and checking the expired time, you can `DELETE FROM http_cache WHERE expiration < now();` (or something like that). You probably want an index on expiration time to make the query efficient.

Reply:
I like this one..!!

Reply:
Fixed